src/include/network/postgres_protocol_utils.h

```diff
@@ -265,14 +313,72 @@ class PostgresPacketWriter {
     BeginPacket(NetworkMessageType::SIMPLE_QUERY_COMMAND).AppendString(query).EndPacket();
   }
 
+  /**
+   * Writes an empty query response
+   */
+  void WriteEmptyQueryResponse() { BeginPacket(NetworkMessageType::EMPTY_QUERY_RESPONSE).EndPacket(); }
+
+  /**
+   * Writes a no-data response
+   */
+  void WriteNoData() { BeginPacket(NetworkMessageType::NO_DATA_RESPONSE).EndPacket(); }
+
+  void WriteParameterDescription(const std::vector<PostgresValueType> &param_types) {
+    BeginPacket(NetworkMessageType::PARAMETER_DESCRIPTION);
+    AppendValue<int16_t>(static_cast<int16_t>(param_types.size()));
+
+    for (auto &type : param_types) AppendValue<int32_t>(static_cast<int32_t>(type));
+
+    EndPacket();
+  }
+
+  /**
+   * Writes row description, as the first packet of sending query results
+   * @param columns the column names
+   */
+  void WriteRowDescription(const std::vector<std::string> &columns) {
+    // TODO(Weichen): fill correct OIDs here. This depends on the catalog.
+    BeginPacket(NetworkMessageType::ROW_DESCRIPTION).AppendValue<int16_t>(static_cast<int16_t>(columns.size()));
+    for (auto &col_name : columns) {
+      AppendString(col_name)
+          .AppendValue<int32_t>(0)                                     // table oid, 0 for now
+          .AppendValue<int16_t>(0)                                     // column oid, 0 for now
+          .AppendValue(static_cast<int32_t>(PostgresValueType::TEXT))  // type oid
+          .AppendValue<int16_t>(-1)                                    // Variable Length
+          .AppendValue<int32_t>(-1)                                    // pg_attribute.attrmod, generally -1
+          .AppendValue<int16_t>(0);                                    // text=0
+    }
+
+    EndPacket();
+  }
+
+  /**
+   * Writes a data row.
+   * @param values a row's values.
+   */
+  void WriteDataRow(const std::vector<std::string> &values) {
```

Comment:
Is the real system going to be passing a vector of strings?

Reply:
I don't think so, the `ResultSet` should have actual column schema rather than all strings... I haven't updated the `ResultSet` yet so it is like that right now.